File: src/lib/quota-windows.ts
export type QuotaWindowStatus = "available" | "near_cap" | "exhausted" | "unknown";

export interface QuotaEventLike {
  provider: string;
  service?: string | null;
  label?: string | null;
  credits?: number | null;
  limit?: number | null;
  occurredAt: Date | string;
  metadata?: unknown;
}

export interface SkipModelType {
  instanceId: string;
  model: string;
}

export interface QuotaWindow {
  id: string;
  provider: string;
  /** Canonical provider key the window is grouped under (see PROVIDER_ALIASES). */
  providerKey: string;
  /** Human label for the provider, e.g. "Claude".  Additive; safe to ignore. */
  providerLabel: string;
  /**
   * Set to "antigravity" when the window comes from Antigravity's own routing
   * buckets rather than the named vendor's subscription.  Antigravity reports a
   * bucket called "Claude and GPT models"; that is NOT the user's Claude plan.
   */
  via: string | null;
  sourceApp: string | null;
  modelId: string | null;
  modelType: string | null;
  label: string;
  remainingPercent: number | null;
  remainingUnknown: boolean;
  isExhausted: boolean;
  resetAt: string | null;
  window: string | null;
  status: QuotaWindowStatus;
  skip: boolean;
  skipReason: string | null;
  occurredAt: string;
  source: string | null;
}

/** One provider's subscription quota windows.  Additive to the v1 response. */
export interface QuotaProviderGroup {
  provider: string;
  providerLabel: string;
  via: string | null;
  /** True for the five providers the dashboard always shows a row for. */
  expected: boolean;
  windows: QuotaWindow[];
}

export interface QuotaWindowsResponse {
  generatedAt: string;
  windows: QuotaWindow[];
  skipModelTypes: SkipModelType[];
  /**
   * Windows grouped by provider, with an entry for every expected provider even
   * when it has reported nothing yet (empty `windows`).  A provider that is
   * missing should be visible, not silently absent.
   */
  providerGroups: QuotaProviderGroup[];
}

const ANTIGRAVITY_INSTANCE = "antigravity";

/**
 * Canonical provider keys for subscription quota reporting, in display order.
 * The dashboard renders a row for every one of these, reported or not.
 */
export const EXPECTED_QUOTA_PROVIDERS = [
  "anthropic",
  "openai",
  "google-antigravity",
  "xai",
  "minimax",
] as const;

/** Event `provider` values that should collapse onto one canonical key. */
const PROVIDER_ALIASES: Record<string, string> = {
  anthropic: "anthropic",
  "claude-code": "anthropic",
  claude: "anthropic",
  openai: "openai",
  "openai-codex": "openai",
  codex: "openai",
  google: "google-antigravity",
  "google-antigravity": "google-antigravity",
  antigravity: "google-antigravity",
  "antigravity-cli": "google-antigravity",
  xai: "xai",
  "grok-build": "xai",
  grok: "xai",
  minimax: "minimax",
  "minimax-code": "minimax",
};

const PROVIDER_LABELS: Record<string, string> = {
  anthropic: "Claude",
  openai: "Codex",
  "google-antigravity": "Antigravity",
  xai: "Grok",
  minimax: "MiniMax",
};

/** Collapse an event `provider` onto the key its windows are grouped under. */
export function quotaProviderKey(provider: string): string {
  const raw = String(provider ?? "").trim().toLowerCase();
  return PROVIDER_ALIASES[raw] ?? raw;
}

/** Human label for a provider key.  Falls back to the raw slug. */
export function quotaProviderLabel(provider: string): string {
  const key = quotaProviderKey(provider);
  return PROVIDER_LABELS[key] ?? (key || "Unknown");
}

/**
 * Antigravity routes to several vendors' models under its own subscription, so
 * its buckets must never be presented as the user's Claude or ChatGPT plan.
 */
export function quotaProviderVia(provider: string): string | null {
  return quotaProviderKey(provider) === "google-antigravity" ? "antigravity" : null;
}

const CLAUDE_GPT_MODELS = [
  "claude-opus-4-6-thinking",
  "claude-sonnet-4-6",
  "gpt-oss-120b-medium",
];

const GEMINI_MODELS = [
  "gemini-3.8-flash-high",
  "gemini-3.8-flash-medium",
  "gemini-3.8-flash-low",
  "gemini-3.7-flash-high",
  "gemini-3.7-flash-medium",
  "gemini-3.7-flash-low",
  "gemini-3.6-flash-high",
  "gemini-3.6-flash-medium",
  "gemini-3.6-flash-low",
  "gemini-3.1-pro-high",
  "gemini-3.1-pro-low",
  "gemini-3-flash",
];

function asRecord(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  return value as Record<string, unknown>;
}

function asString(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

function asBoolean(value: unknown): boolean {
  return value === true;
}

function iso(value: Date | string): string {
  if (value instanceof Date) return value.toISOString();
  const parsed = Date.parse(value);
  return Number.isFinite(parsed) ? new Date(parsed).toISOString() : String(value);
}

export function quotaStatus(input: {
  remainingPercent: number | null;
  remainingUnknown: boolean;
  isExhausted: boolean;
}): QuotaWindowStatus {
  if (input.isExhausted || (input.remainingPercent != null && input.remainingPercent <= 0)) {
    return "exhausted";
  }
  // antigravity-usage prints N/A when remainingPercentage is omitted.
  // Owner 2026-09-04: that means none remains.
  if (input.remainingUnknown || input.remainingPercent == null) return "exhausted";
  if (input.remainingPercent < 20) return "near_cap";
  return "available";
}

function skipTargetsFor(window: QuotaWindow): SkipModelType[] {
  if (!window.skip) return [];
  // skipModelTypes drives Antigravity instance routing only.  Claude/Codex/
  // Grok/MiniMax subscription windows must never emit an antigravity skip.
  if (window.via !== "antigravity") return [];
  if (window.modelId) {
    return [{ instanceId: ANTIGRAVITY_INSTANCE, model: window.modelId }];
  }
  const group = `${window.label} ${window.provider}`.toLowerCase();
  if (group.includes("claude") || group.includes("gpt")) {
    return CLAUDE_GPT_MODELS.map((model) => ({ instanceId: ANTIGRAVITY_INSTANCE, model }));
  }
  if (group.includes("gemini")) {
    return GEMINI_MODELS.map((model) => ({ instanceId: ANTIGRAVITY_INSTANCE, model }));
  }
  return [];
}

export function projectQuotaWindows(
  events: QuotaEventLike[],
  now = new Date(),
): QuotaWindowsResponse {
  const latest = new Map<string, QuotaWindow>();
  for (const event of events) {
    const meta = asRecord(event.metadata);
    const modelId = asString(meta.modelId);
    const bucketId = asString(meta.bucketId);
    const series = modelId ?? bucketId ?? `${event.provider}:${event.label ?? ""}`;
    if (latest.has(series)) continue;

    const limit = typeof event.limit === "number" && event.limit > 0 ? event.limit : 100;
    const omitted = asBoolean(meta.remainingUnknown) || event.credits == null;
    const remainingPercent = omitted
      ? 0
      : Math.round((Number(event.credits) / limit) * 10_000) / 100;
    const isExhausted =
      asBoolean(meta.isExhausted) || omitted || remainingPercent <= 0;
    const remainingUnknown = false;
    const status = quotaStatus({ remainingPercent, remainingUnknown, isExhausted });
    latest.set(series, {
      id: series,
      provider: event.provider,
      providerKey: quotaProviderKey(event.provider),
      providerLabel: quotaProviderLabel(event.provider),
      via: quotaProviderVia(event.provider),
      sourceApp: event.service ?? null,
      modelId,
      modelType: modelId,
      label: event.label ?? modelId ?? event.provider,
      remainingPercent,
      remainingUnknown,
      isExhausted,
      resetAt: asString(meta.resetAt),
      window: asString(meta.quotaWindow),
      status,
      skip: status === "exhausted",
      skipReason:
        status === "exhausted"
          ? `${event.label ?? modelId ?? "model"} remaining ${remainingPercent ?? 0}%`
          : null,
      occurredAt: iso(event.occurredAt),
      source: asString(meta.source),
    });
  }

  const windows = [...latest.values()];
  const skipModelTypes: SkipModelType[] = [];
  const seenSkip = new Set<string>();
  for (const window of windows) {
    for (const target of skipTargetsFor(window)) {
      const key = `${target.instanceId}:${target.model}`;
      if (seenSkip.has(key)) continue;
      seenSkip.add(key);
      skipModelTypes.push(target);
    }
  }

  return {
    generatedAt: now.toISOString(),
    windows,
    skipModelTypes,
    providerGroups: groupWindowsByProvider(windows),
  };
}

/**
 * Group windows by canonical provider.  Every expected provider gets a group
 * even with no windows, so the dashboard can show "no quota report yet" instead
 * of quietly omitting the provider.
 */
export function groupWindowsByProvider(windows: QuotaWindow[]): QuotaProviderGroup[] {
  const groups = new Map<string, QuotaProviderGroup>();
  for (const provider of EXPECTED_QUOTA_PROVIDERS) {
    groups.set(provider, {
      provider,
      providerLabel: quotaProviderLabel(provider),
      via: quotaProviderVia(provider),
      expected: true,
      windows: [],
    });
  }
  for (const window of windows) {
    const key = window.providerKey;
    let group = groups.get(key);
    if (!group) {
      group = {
        provider: key,
        providerLabel: quotaProviderLabel(key),
        via: quotaProviderVia(key),
        expected: false,
        windows: [],
      };
      groups.set(key, group);
    }
    group.windows.push(window);
  }
  for (const group of groups.values()) {
    group.windows.sort((a, b) => a.label.localeCompare(b.label));
  }
  return [...groups.values()];
}

File: ios/UsageMonitor/UsageMonitorKit/Sources/Models/QuotaWindows.swift
import Foundation

// ---------------------------------------------------------------------------
// `GET /api/quota-windows` — the latest remaining-percent quota windows the
// server tracks per provider (5h / 7d / 24h / monthly cadences). The route
// was originally built for BotFleet skip-model routing; this is the first
// iOS consumer that surfaces it to a human so they can see % remaining or
// used on their subscriptions (Claude, Codex, Antigravity, Grok, MiniMax).
//
// Decoding follows the house rule for server payloads (see
// `PlatformStatus.swift`): every field is optional or defaulted and unknown
// enum values fall back to a safe case, so a server-side shape change
// degrades one row instead of failing the whole card. A parallel server
// change (owner 2026-09-12) is landing Mac collectors for Claude, Codex,
// Grok, and MiniMax that emit this same shape, plus two ADDITIVE fields — a
// human provider label and a `via: "antigravity"` marker on Antigravity's
// internal model-family buckets. Unknown keys are simply ignored by
// `Decodable`; the two new fields below decode as `nil` until the server
// change ships.
// ---------------------------------------------------------------------------

public struct QuotaWindowsResponse: Codable, Hashable, Sendable {
    public var ok: Bool
    public var generatedAt: String?
    public var windows: [QuotaWindow]

    public var generatedAtDate: Date? {
        generatedAt.flatMap { ISO8601DateParser.date(from: $0) }
    }

    public init(ok: Bool = true, generatedAt: String? = nil, windows: [QuotaWindow] = []) {
        self.ok = ok
        self.generatedAt = generatedAt
        self.windows = windows
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        ok = (try? container.decode(Bool.self, forKey: .ok)) ?? true
        generatedAt = try? container.decodeIfPresent(String.self, forKey: .generatedAt)
        windows = (try? container.decode([QuotaWindow].self, forKey: .windows)) ?? []
    }
}

/// Matches the server's `QuotaWindowStatus` union in `src/lib/quota-windows.ts`.
public enum QuotaWindowStatus: String, Codable, Hashable, Sendable {
    case available
    case nearCap = "near_cap"
    case exhausted
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = (try? decoder.singleValueContainer().decode(String.self)) ?? ""
        self = QuotaWindowStatus(rawValue: raw) ?? .unknown
    }
}

public struct QuotaWindow: Codable, Hashable, Sendable, Identifiable {
    public var id: String
    public var provider: String
    public var sourceApp: String?
    public var modelId: String?
    public var modelType: String?
    public var label: String
    public var remainingPercent: Double?
    public var remainingUnknown: Bool
    public var isExhausted: Bool
    public var resetAt: String?
    public var window: String?
    public var status: QuotaWindowStatus
    public var occurredAt: String?
    public var source: String?
    /// Additive (owner 2026-09-12): a human-facing provider label, when the
    /// server has one nicer than the raw `provider` slug.
    public var providerLabel: String?
    /// Additive (owner 2026-09-12): set on Antigravity's internal
    /// model-family buckets so the UI can caption them "via Antigravity"
    /// even if they're later reclassified under the model's own provider.
    public var via: String?

    public var resetAtDate: Date? {
        resetAt.flatMap { ISO8601DateParser.date(from: $0) }
    }

    /// The value to render: the reported percent, or 0 when the server marked
    /// this window exhausted/unknown without a number. Never invented.
    public var displayRemainingPercent: Double {
        remainingPercent ?? 0
    }

    public init(
        id: String,
        provider: String,
        sourceApp: String? = nil,
        modelId: String? = nil,
        modelType: String? = nil,
        label: String,
        remainingPercent: Double? = nil,
        remainingUnknown: Bool = false,
        isExhausted: Bool = false,
        resetAt: String? = nil,
        window: String? = nil,
        status: QuotaWindowStatus = .unknown,
        occurredAt: String? = nil,
        source: String? = nil,
        providerLabel: String? = nil,
        via: String? = nil
    ) {
        self.id = id
        self.provider = provider
        self.sourceApp = sourceApp
        self.modelId = modelId
        self.modelType = modelType
        self.label = label
        self.remainingPercent = remainingPercent
        self.remainingUnknown = remainingUnknown
        self.isExhausted = isExhausted
        self.resetAt = resetAt
        self.window = window
        self.status = status
        self.occurredAt = occurredAt
        self.source = source
        self.providerLabel = providerLabel
        self.via = via
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = (try? container.decode(String.self, forKey: .id)) ?? UUID().uuidString
        provider = (try? container.decode(String.self, forKey: .provider)) ?? "unknown"
        sourceApp = try? container.decodeIfPresent(String.self, forKey: .sourceApp)
        modelId = try? container.decodeIfPresent(String.self, forKey: .modelId)
        modelType = try? container.decodeIfPresent(String.self, forKey: .modelType)
        label = (try? container.decode(String.self, forKey: .label)) ?? provider
        remainingPercent = try? container.decodeIfPresent(Double.self, forKey: .remainingPercent)
        remainingUnknown = (try? container.decode(Bool.self, forKey: .remainingUnknown)) ?? false
        isExhausted = (try? container.decode(Bool.self, forKey: .isExhausted)) ?? false
        resetAt = try? container.decodeIfPresent(String.self, forKey: .resetAt)
        window = try? container.decodeIfPresent(String.self, forKey: .window)
        status = (try? container.decode(QuotaWindowStatus.self, forKey: .status)) ?? .unknown
        occurredAt = try? container.decodeIfPresent(String.self, forKey: .occurredAt)
        source = try? container.decodeIfPresent(String.self, forKey: .source)
        providerLabel = try? container.decodeIfPresent(String.self, forKey: .providerLabel)
        via = try? container.decodeIfPresent(String.self, forKey: .via)
    }
}

/// The five subscription providers the Overview card always accounts for,
/// even when the server has nothing to say about one of them yet.
public enum SubscriptionQuotaProvider: String, CaseIterable, Hashable, Sendable {
    case claude
    case codex
    case antigravity
    case grok
    case minimax

    public var displayName: String {
        switch self {
        case .claude: return "Claude"
        case .codex: return "Codex"
        case .antigravity: return "Antigravity"
        case .grok: return "Grok"
        case .minimax: return "MiniMax"
        }
    }

    /// Classify a raw server `provider` slug (e.g. `"google-antigravity"`)
    /// into one of the five expected subscription providers. Keyword-based
    /// rather than exact-match: the Mac collectors for Claude/Codex/Grok/
    /// MiniMax are landing in a parallel change and may emit a slightly
    /// different slug than assumed here. A provider that matches nothing
    /// is simply not shown — never invented as one of the five.
    public static func classify(_ raw: String) -> SubscriptionQuotaProvider? {
        let slug = raw.lowercased()
        if slug.contains("antigrav") { return .antigravity }
        if slug.contains("codex") { return .codex }
        if slug.contains("claude") || slug.contains("anthropic") { return .claude }
        if slug.contains("grok") || slug.contains("xai") { return .grok }
        if slug.contains("minimax") || slug.contains("mmx") { return .minimax }
        return nil
    }
}

/// "Resets in 3h 24m" style countdown, shared by the card and its tests.
public enum QuotaCountdownFormat {
    public static func label(resetAt: Date?, now: Date = Date()) -> String {
        guard let resetAt else { return "Reset time unknown" }
        let seconds = resetAt.timeIntervalSince(now)
        if seconds <= 0 { return "Resetting now" }
        let totalMinutes = Int(seconds / 60)
        let days = totalMinutes / 1_440
        let hours = (totalMinutes % 1_440) / 60
        let minutes = totalMinutes % 60

        if days > 0 { return "Resets in \(days)d \(hours)h" }
        if hours > 0 { return "Resets in \(hours)h \(minutes)m" }
        if minutes > 0 { return "Resets in \(minutes)m" }
        return "Resets in <1m"
    }
}

/// One provider's slot in the Subscription Quotas card. Always present for
/// all five `SubscriptionQuotaProvider` cases — `windows` is empty rather
/// than the section being omitted, so the card can show an honest "No quota
/// report yet" row per the house rule against inventing data.
public struct SubscriptionQuotaSection: Identifiable, Hashable, Sendable {
    public var provider: SubscriptionQuotaProvider
    /// Sorted lowest-remaining-first, then by label.
    public var windows: [QuotaWindow]

    public var id: SubscriptionQuotaProvider { provider }

    /// `nil` when this provider has no windows yet — sorts after every
    /// provider that does have data.
    public var minRemainingPercent: Double? {
        windows.map(\.displayRemainingPercent).min()
    }

    public init(provider: SubscriptionQuotaProvider, windows: [QuotaWindow]) {
        self.provider = provider
        self.windows = windows
    }
}

public extension QuotaWindowsResponse {
    /// Groups `windows` into one section per expected subscription provider,
    /// ordered by lowest remaining percent first (most urgent on top), then
    /// by provider name. Providers the server hasn't reported yet still
    /// appear, with an empty `windows` array, so the card renders an honest
    /// empty row instead of silently dropping a provider the owner expects
    /// to see (Claude, Codex, Antigravity, Grok, MiniMax).
    var subscriptionSections: [SubscriptionQuotaSection] {
        var buckets: [SubscriptionQuotaProvider: [QuotaWindow]] = [:]
        for window in windows {
            guard let provider = SubscriptionQuotaProvider.classify(window.provider) else { continue }
            buckets[provider, default: []].append(window)
        }

        let sections = SubscriptionQuotaProvider.allCases.map { provider -> SubscriptionQuotaSection in
            let sortedWindows = (buckets[provider] ?? []).sorted { lhs, rhs in
                if lhs.displayRemainingPercent != rhs.displayRemainingPercent {
                    return lhs.displayRemainingPercent < rhs.displayRemainingPercent
                }
                return lhs.label < rhs.label
            }
            return SubscriptionQuotaSection(provider: provider, windows: sortedWindows)
        }

        return sections.sorted { lhs, rhs in
            let lhsKey = lhs.minRemainingPercent ?? .infinity
            let rhsKey = rhs.minRemainingPercent ?? .infinity
            if lhsKey != rhsKey { return lhsKey < rhsKey }
            return lhs.provider.displayName < rhs.provider.displayName
        }
    }
}
